tests/unit/Alignment.cpp

```diff
@@ -0,0 +1,240 @@
+  return Alignment != 0 ? std::to_string(Alignment) : "undefined";
+}
+
+template<std::size_t Count>
+struct ABIList {
+  std::array<const abi::Definition *, Count> Values;
+
+  // clang-format off
+  template<auto... ExpectedValues>
+    requires((convertible_to<decltype(ExpectedValues), std::uint64_t> && ...)
+              && sizeof...(ExpectedValues) == Count)
+  void test(const model::QualifiedType &Type) {
+    // clang-format on
+    constexpr std::array<std::uint64_t, Count> Expected = { ExpectedValues... };
+    constexprRepeat<Count>([&Type, &Expected, this]<std::size_t Index>() {
+      std::optional<std::uint64_t> TestResult = Values[Index]->alignment(Type);
+      if (TestResult.value_or(0) != Expected[Index]) {
+        std::string Error = "Alignment test failed for type:\n"
+                            + serializeToString(Type) + "ABI ('"
+                            + serializeToString(Values[Index]->ABI())
+                            + "') reports the alignment of '"
+                            + printAlignment(TestResult.value_or(0)) + "', "
+                            + "while the expected value is '"
+                            + printAlignment(Expected[Index]) + "'.\n";
+        revng_abort(Error.c_str());
+      }
+    });
+  }
+
+  void test(const model::QualifiedType &LHS, const model::QualifiedType &RHS) {
+    constexprRepeat<Count>([&LHS, &RHS, this]<std::size_t Index>() {
+      std::optional<std::uint64_t> Left = Values[Index]->alignment(LHS);
+      std::optional<std::uint64_t> Right = Values[Index]->alignment(RHS);
+      if (Left != Right) {
+        std::string Error = "Alignment comparison test failed for types:\n"
+                            + serializeToString(LHS) + "and\n"
+                            + serializeToString(RHS) + "ABI ('"
+                            + serializeToString(Values[Index]->ABI())
+                            + "') reports the alignment of '"
+                            + printAlignment(Left.value_or(0))
+                            + "' for the first one, and '"
+                            + printAlignment(Right.value_or(0))
+                            + "' for the second one.\n";
+        revng_abort(Error.c_str());
+      }
+    });
+  }
+};
+
+// clang-format off
+template<auto... ABINames>
+  requires(std::same_as<decltype(ABINames), model::ABI::Values> && ...)
+static ABIList<sizeof...(ABINames)> prepareABIs() {
+  // clang-format on
+  return { std::array<const abi::Definition *, sizeof...(ABINames)>{
+    &abi::Definition::get(ABINames)... } };
+}
+
+static model::QualifiedType makePrimitive(model::PrimitiveTypeKind::Values Kind,
+                                          std::size_t Size,
+                                          model::Binary &Binary) {
+  return model::QualifiedType(Binary.getPrimitiveType(Kind, Size), {});
+}
+
+BOOST_AUTO_TEST_CASE(PrimitiveTypes) {
+  ABIList ABIs = prepareABIs<model::ABI::AAPCS64,
```

Comment:
Template usage is extremely obscure and brings no benefits.

Make a vector of inputs with expected outputs and write a simple function iterating over it and checking expectations.

```c++
std::vector<TestData> Tests = {
  { .ABI = ..., .Type = ..., .ExpectedAlignment = ... }
};
```

Do not use templates at all, each and every concept you're dealing with here is a runtime concept, there's no need for templates.

Reply:
What about now?

Reply:
What about now?

Reply:
I don't see a data structure with inputs and expected results, but maybe it's better to close off the other discussion first.

In general my idea is to have tests organized in a way that's easy to turn them into command line tools that accept well defined inputs and emit well defined outputs.

Let's not do this now, but in general it's better to pivot your code from having a lot of code to having a lot of data and a smart interpreter of the data.

Reply:
The other thread is closed. What should we do about this test then?

Reply:
What we have is OK, except for the other comment.